Sources/Logging/Logging.swift

```diff
@@ -662,41 +683,146 @@ public enum LoggingSystem {
+    internal static var metadataProvider: Logger.MetadataProvider? {
+        return self._metadataProviderFactory.metadataProvider
     }
 
     private final class FactoryBox {
         private let lock = ReadWriteLock()
-        fileprivate var _underlying: (String) -> LogHandler
+        fileprivate var _underlying: (_ label: String, _ provider: Logger.MetadataProvider?) -> LogHandler
         private var initialized = false
 
-        init(_ underlying: @escaping (String) -> LogHandler) {
+        init(_ underlying: @escaping (String, Logger.MetadataProvider?) -> LogHandler) {
             self._underlying = underlying
         }
 
-        func replaceFactory(_ factory: @escaping (String) -> LogHandler, validate: Bool) {
+        func replaceFactory(_ factory: @escaping (String, Logger.MetadataProvider?) -> LogHandler, validate: Bool) {
             self.lock.withWriterLock {
                 precondition(!validate || !self.initialized, "logging system can only be initialized once per process.")
                 self._underlying = factory
                 self.initialized = true
             }
         }
 
-        var underlying: (String) -> LogHandler {
+        var underlying: (String, Logger.MetadataProvider?) -> LogHandler {
             return self.lock.withReaderLock {
                 return self._underlying
             }
         }
     }
+
+    private final class MetadataProviderBox {
+        private let lock = ReadWriteLock()
+
+        internal var _underlying: Logger.MetadataProvider?
+        private var initialized = false
+
+        init(_ underlying: Logger.MetadataProvider?) {
+            self._underlying = underlying
+        }
+
+        func replaceMetadataProvider(_ metadataProvider: Logger.MetadataProvider?, validate: Bool) {
+            self.lock.withWriterLock {
+                precondition(!validate || !self.initialized, "logging system can only be initialized once per process.")
+                self._underlying = metadataProvider
+                self.initialized = true
+            }
+        }
+
+        var metadataProvider: Logger.MetadataProvider? {
+            return self.lock.withReaderLock {
+                return self._underlying
+            }
+        }
+    }
+}
+
+extension LoggingSystem {
+    /// `bootstrap` is a one-time configuration function which globally selects the desired logging backend
+    /// implementation. 
+    ///
+    /// - Warning:
+    /// `bootstrap` can be called at maximum once in any given program, calling it more than once will
+    /// lead to undefined behavior, most likely a crash.
+    ///
+    /// - parameters:
+    ///     - metadataProvider: The `MetadataProvider` used to inject runtime-generated metadata, defaults to a "no-op" provider.
+    ///     - factory: A closure that given a `Logger` identifier, produces an instance of the `LogHandler`.
+    public static func bootstrap(metadataProvider: Logger.MetadataProvider?,
```

Comment:
metadataProvider should be second argument since optional?

Reply:
right, thanks :)

Reply:
Heh, so I did this, but the callsites become very ugly...

```
LoggingSystem.bootstrapInternal() { label, metadataProvider in
  logging.makeWithMetadataProvider(label: label, metadataProvider: metadataProvider)
}
LoggingSystem.bootstrapInternal(metadataProvider: .myCoolTracer) { label, metadataProvider in
  logging.makeWithMetadataProvider(label: label, metadataProvider: metadataProvider)
}
```

vs

```
LoggingSystem.bootstrapInternal() { label, metadataProvider in
  logging.makeWithMetadataProvider(label: label, metadataProvider: metadataProvider)
}

LoggingSystem.bootstrapInternal({ label, metadataProvider in
logging.makeWithMetadataProvider(label: label, metadataProvider: metadataProvider)
}, metadataProvider: .myCoolTracer)
```

So... I'd suggest to keep existing API but we can change if you feel strongly about it 🤔